libpromises/evalfunction.c

```diff
@@ -4670,14 +4680,36 @@ static FnCallResult FnCallIPRange(EvalContext *ctx, ARG_UNUSED const Policy *pol
 
     for (const Item *ip = EvalContextGetIpAddresses(ctx); ip != NULL; ip = ip->next)
     {
-        if (FuzzySetMatch(range, VIPADDRESS) == 0 ||
-            FuzzySetMatch(range, ip->name) == 0)
+        Rlist *ifaces = finalargs->next;
+        // we match on VIPADDRESS iff no interfaces were requested
+        if (FuzzySetMatch(range, VIPADDRESS) == 0 && NULL == ifaces)
         {
-            return FnReturnContext(true);
+            Log(LOG_LEVEL_DEBUG, "%s: found range %s on hostip %s, no interface", fp->name, range, ip->name);
+            return FnReturnContextTrue();
+        }
+        else if (FuzzySetMatch(range, ip->name) == 0)
+        {
+            if (NULL == ifaces) // no interfaces requested
+            {
+                Log(LOG_LEVEL_DEBUG, "%s: found range %s on IP %s, any interface", fp->name, range, ip->name);
+                return FnReturnContextTrue();
+            }
+            else // check the specific interfaces given as arguments
+            {
+                for (; ifaces != NULL; ifaces = ifaces->next)
+                {
+                    Log(LOG_LEVEL_DEBUG, "%s: checking range %s on IP %s, interface %s", fp->name, range, ip->name, RlistScalarValue(ifaces));
+                    if (0 == strcmp(RlistScalarValue(ifaces), ip->classes))
```

Comment:
The reason it fails on Solaris is because the interface is named `e1000g0:4`. However, the interface name is canonified in `GetInterfacesInfo()`, but not here, leading to a false comparison. I'd fix it myself, but I'm not sure what is the right approach.

The Most Correct (tm) way is to not canonify in the first place, since this is an interface name and not a class name. However, you're (slightly) abusing the classes field of the `Item` object here to hold the interface name in a class field. Do you think there will be any ill effects of not canonifying it? IOW are we using it in a class context anywhere? I suppose we shouldn't be, since it used to be empty.

Linux can also have `eth0:1` type interfaces, so this is important there as well.